File: sandbox-runner/patterns/delayed-retry/java/src/main/java/Starter.java
import io.temporal.client.WorkflowClient;
import io.temporal.client.WorkflowOptions;
import io.temporal.serviceclient.WorkflowServiceStubs;

public class Starter {
    public static void main(String[] args) {
        WorkflowServiceStubs service = WorkflowServiceStubs.newLocalServiceStubs();
        WorkflowClient client = WorkflowClient.newInstance(service);

        String workflowId = Shared.WORKFLOW_ID_PREFIX + "-" + System.currentTimeMillis();

        System.out.println("Starting rate-limited API workflow.");
        System.out.println("Attempt 1 will receive a simulated HTTP 429 with Retry-After: 5s.");
        System.out.println(
                "The nextRetryDelay overrides the RetryPolicy interval for that single retry.");
        System.out.println();

        RateLimitedWorkflow workflow = client.newWorkflowStub(
                RateLimitedWorkflow.class,
                WorkflowOptions.newBuilder()
                        .setTaskQueue(Shared.TASK_QUEUE)
                        .setWorkflowId(workflowId)
                        .build());

        System.out.println("Started workflow: " + workflowId);

        String result = workflow.run("https://api.example.com/data");
        System.out.println("Workflow result: " + result);
        System.out.println();
        System.out.println("Open the Temporal UI and search for '" + workflowId
                + "' to see the retry timing.");
    }
}

File: sandbox-runner/patterns/delayed-retry/typescript/starter.ts
import { Client, Connection } from "@temporalio/client";

import { TASK_QUEUE, WORKFLOW_ID_PREFIX } from "./shared";
import { rateLimitedWorkflow } from "./workflows";

async function main(): Promise<void> {
  const connection = await Connection.connect();
  try {
    const client = new Client({ connection });
    const runId = `${WORKFLOW_ID_PREFIX}-${Date.now()}`;

    console.log("Starting rate-limited API workflow.");
    console.log(
      "Attempt 1 will receive a simulated HTTP 429 with Retry-After: 5s.",
    );
    console.log(
      "The nextRetryDelay overrides the RetryPolicy interval for that single retry.",
    );
    console.log();

    const handle = await client.workflow.start(rateLimitedWorkflow, {
      args: ["https://api.example.com/data"],
      taskQueue: TASK_QUEUE,
      workflowId: runId,
    });
    console.log(`Started workflow: ${handle.workflowId}`);

    const result = await handle.result();
    console.log(`Workflow result: ${result}`);
    console.log();
    console.log(
      `Open the Temporal UI and search for '${handle.workflowId}' to see the retry timing.`,
    );
  } finally {
    await connection.close();
  }
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
